src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -651,7 +651,33 @@ private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ImmutableArray<
 
             var rootDecisionDagNode = decisionDag.RootNode.Dag;
             RoslynDebug.Assert(rootDecisionDagNode != null);
-            return new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
+            var boundDecisionDag = new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
+#if DEBUG
+            int nextTempNumber = 0;
+            // Note that this leverages the `BoundDagEvaluation.GetHashCode()`
+            // to make "equivalent" evaluation nodes share an ID
+            var tempIdentifierMap = PooledDictionary<BoundDagEvaluation, int>.GetInstance();
+            int tempIdentifier(BoundDagEvaluation e)
+            {
+                return tempIdentifierMap.TryGetValue(e, out int value) ? value : tempIdentifierMap[e] = ++nextTempNumber;
+            }
+
+            var sortedBoundDagNodes = boundDecisionDag.TopologicallySortedNodes;
+            for (int i = 0; i < sortedBoundDagNodes.Length; i++)
+            {
+                sortedBoundDagNodes[i].Id = i;
+                if (sortedBoundDagNodes[i] is BoundEvaluationDecisionDagNode { Evaluation: var evaluation })
```

Comment:
Consider commenting on why only evaluations get this treatment (it's where new temp/values come from).
Note: we may need to give the initial value an explicit identifier `0` if the unitialized value is changed to `-1`.

Reply:
Done